File: sdk/scheduler.ts
import { canonicalJson, type Json } from "./json.ts";
import { collection, component, fail, methodInfo, plainObject, requireName, task } from "./core.ts";
import { ValidationError } from "./schema.ts";
import type { ArgsOf, ArgsParameter, Collection, Component, Context, Failure, MutationContext, MutationMethod, TaskFailure } from "./core.ts";

export interface ScheduledTimer<A = Json> {
  state: "pending" | "failed";
  handler: string;
  args: A;
  dueAt: number;
  attempts: number;
  error: Failure | null;
  createdAt: number;
  updatedAt: number;
}
export interface Timer<A = Json> extends ScheduledTimer<A> { id: string }

export interface SchedulerOptions {
  /** Failed attempts before a timer stays failed. Default 3. */
  readonly maxAttempts?: number;
  /** First retry delay in milliseconds, doubling per failure. Default 1000. */
  readonly retryDelayMs?: number;
  /** Upper bound on retry delays. Default 60000. */
  readonly maxRetryDelayMs?: number;
}

type Handlers = Readonly<Record<string, MutationMethod<any, any>>>;

export interface Scheduler<R extends Handlers> extends Component {
  readonly records: Collection<ScheduledTimer, string, { readonly due: readonly ["state", "dueAt"] }>;
  /** Replacing an ID debounces earlier work and starts a new retry budget. */
  after<H extends Extract<keyof R, string>>(ctx: MutationContext, id: string, delayMs: number, handler: H, ...args: ArgsParameter<ArgsOf<R[H]>>): Timer<ArgsOf<R[H]>>;
  at<H extends Extract<keyof R, string>>(ctx: MutationContext, id: string, dueAt: number, handler: H, ...args: ArgsParameter<ArgsOf<R[H]>>): Timer<ArgsOf<R[H]>>;
  cancel(ctx: MutationContext, id: string): boolean;
  get(ctx: Context, id: string): Timer | null;
  /** Timers in deadline order, optionally one state. */
  scan(ctx: Context, options?: { readonly state?: "pending" | "failed" }): Timer[];
  /** Requeue a failed timer with the deployed handler and a fresh retry budget. */
  retry(ctx: MutationContext, id: string, delayMs?: number): Timer;
}

function integer(value: unknown, label: string, minimum = 0): asserts value is number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < minimum) {
    throw new TypeError(`${label} must be a safe integer at least ${minimum}`);
  }
}

function clock(ctx: Context): number {
  const now = ctx.now();
  integer(now, "Server time");
  return now;
}

/**
 * Durable callbacks: mutation handlers run after a deadline in their own
 * transaction. Add the scheduler to define({ uses }). Deadlines mean "not
 * before"; a handler's writes and the timer's removal commit together.
 */
export function scheduler<const R extends Handlers>(name: string, handlers: R, options: SchedulerOptions = {}): Scheduler<R> {
  requireName(name, "Scheduler name");
  if (name.startsWith("$flower.")) throw new TypeError("Scheduler names beginning with $flower. are reserved");
  const settings = plainObject(options, "Scheduler options", ["maxAttempts", "retryDelayMs", "maxRetryDelayMs"]);
  const maxAttempts = settings.maxAttempts ?? 3, retryDelayMs = settings.retryDelayMs ?? 1_000, maxRetryDelayMs = settings.maxRetryDelayMs ?? 60_000;
  integer(maxAttempts, "maxAttempts", 1);
  integer(retryDelayMs, "retryDelayMs", 1);
  integer(maxRetryDelayMs, "maxRetryDelayMs", 1);
  if (maxRetryDelayMs < retryDelayMs) throw new RangeError("Maximum retry delay must be at least the first retry delay");
  const registry: Record<string, MutationMethod<any, any>> = Object.create(null);
  for (const [alias, handler] of Object.entries(plainObject(handlers, "Scheduler handlers"))) {
    requireName(alias, "Handler alias");
    if ((handler as MutationMethod)?.kind !== "mutationMethod" || typeof (handler as MutationMethod).compute !== "function") {
      throw new TypeError("Scheduler handlers must be mutation methods");
    }
    registry[alias] = handler as MutationMethod;
  }
  const records = collection<ScheduledTimer>(name).index("due", ["state", "dueAt"]);

  function handlerFor(alias: string): MutationMethod {
    if (!Object.hasOwn(registry, alias)) fail("SCHEDULER_HANDLER_MISSING", `Unknown scheduler handler ${JSON.stringify(alias)}`);
    return registry[alias];
  }
  function get(ctx: Context, id: string): Timer | null {
    requireName(id, "Timer ID");
    const timer = ctx.get(records, id);
    return timer === null ? null : { id, ...timer };
  }
  function first(ctx: Context, bound: number | null): Timer | null {
    const row = ctx.range(records.by("due").range({ prefix: ["pending"], ...(bound === null ? {} : { lte: bound }), limit: 1 })).rows[0];
    return row ? { id: row.key, ...row.value } : null;
  }
  function schedule(ctx: MutationContext, id: string, dueAt: number, handler: string, args: unknown): Timer {
    requireName(id, "Timer ID");
    integer(dueAt, "Deadline");
    const schema = methodInfo(handlerFor(handler)).args;
    if (schema) {
      try { schema.parse(args ?? null); }
      catch (error) {
        if (error instanceof ValidationError) fail("INVALID_ARGUMENT", `${handler} arguments: ${error.message}`, { path: [...error.path] });
        throw error;
      }
    }
    const time = clock(ctx);
    const previous = ctx.get(records, id);
    const timer: ScheduledTimer = {
      state: "pending", handler, args: JSON.parse(canonicalJson(args ?? null)) as Json, dueAt, attempts: 0, error: null,
      createdAt: previous?.createdAt ?? time, updatedAt: time,
    };
    ctx.set(records, id, timer);
    return { id, ...timer };
  }

  const timers = task(`scheduler:${name}`, {
    due: (ctx) => first(ctx, null)?.dueAt ?? null,
    run(ctx) {
      const due = first(ctx, clock(ctx));
      if (!due) return null;
      // Removal is staged before dispatch; a handler may replace its own ID.
      ctx.delete(records, due.id);
      const result: unknown = handlerFor(due.handler).compute(ctx, due.args);
      canonicalJson(result);
      return { timer: due.id };
    },
    onError(ctx, failure: TaskFailure) {
      const due = first(ctx, clock(ctx));
      if (!due) return null;
      const attempts = due.attempts + 1;
      let delay = retryDelayMs;
      for (let attempt = 1; attempt < attempts && delay < maxRetryDelayMs; attempt++) delay = Math.min(maxRetryDelayMs, delay * 2);
      const exhausted = attempts >= maxAttempts || failure.failedAt > Number.MAX_SAFE_INTEGER - delay;
      const { id, ...previous } = due;
      ctx.set(records, id, {
        ...previous, state: exhausted ? "failed" : "pending", attempts,
        dueAt: exhausted ? due.dueAt : failure.failedAt + delay, updatedAt: failure.failedAt,
        error: { code: failure.error.code, message: failure.error.message, ...(failure.error.details === undefined ? {} : { details: failure.error.details }) },
      });
      return { timer: id, state: exhausted ? "failed" : "pending", attempts };
    },
  });

  return Object.freeze({
    ...component({ collections: [records], tasks: [timers] }),
    records,
    get,
    scan(ctx: Context, filter: { readonly state?: "pending" | "failed" } = {}): Timer[] {
      const { state } = plainObject(filter, "Timer filter", ["state"]);
      if (state !== undefined && state !== "pending" && state !== "failed") throw new TypeError("Unknown timer state filter");
      const timers: Timer[] = [];
      for (const current of (state === undefined ? ["pending", "failed"] : [state]) as ("pending" | "failed")[]) {
        let after: string | undefined;
        do {
          const page = ctx.range(records.by("due").range({ prefix: [current], limit: 64, ...(after === undefined ? {} : { after }) }));
          for (const row of page.rows) timers.push({ id: row.key, ...row.value });
          after = page.cursor ?? undefined;
        } while (after !== undefined);
      }
      return timers.sort((a, b) => a.dueAt - b.dueAt || (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
    },
    after(ctx: MutationContext, id: string, delayMs: number, handler: string, args?: unknown) {
      integer(delayMs, "Delay");
      return schedule(ctx, id, clock(ctx) + delayMs, handler, args);
    },
    at(ctx: MutationContext, id: string, dueAt: number, handler: string, args?: unknown) {
      return schedule(ctx, id, dueAt, handler, args);
    },
    cancel(ctx: MutationContext, id: string): boolean {
      if (get(ctx, id) === null) return false;
      ctx.delete(records, id);
      return true;
    },
    retry(ctx: MutationContext, id: string, delayMs = 0): Timer {
      integer(delayMs, "Delay");
      const timer = get(ctx, id);
      if (timer === null || timer.state !== "failed") fail("TIMER_NOT_FAILED", "Only failed timers can be retried");
      handlerFor(timer.handler);
      const time = clock(ctx);
      const { id: _id, ...stored } = timer;
      const retried: ScheduledTimer = { ...stored, state: "pending", attempts: 0, error: null, dueAt: time + delayMs, updatedAt: time };
      ctx.set(records, id, retried);
      return { id, ...retried };
    },
  }) as unknown as Scheduler<R>;
}

File: crates/flower-sdk/src/scheduler.rs
//! Durable callbacks (the TypeScript SDK's `scheduler`): mutation handlers run
//! after a deadline in their own transaction. Deadlines mean "not before"; a
//! handler's writes and the timer's removal commit together.
use crate::{
    Collection, Ctx, IndexDef, Range, Result, Value,
    app::{Component, Method, Task, TaskFailure},
    fail, fail_with, json, object,
    schema::is_safe_integer,
    type_error,
};
use alloc::{format, string::String, vec, vec::Vec};

const INDEXES: &[IndexDef] = &[IndexDef {
    name: "due",
    fields: &["state", "dueAt"],
}];
const MAX_SAFE_INTEGER: f64 = 9_007_199_254_740_991.0;

pub struct Scheduler {
    pub name: &'static str,
    /// Timers by ID: `{state, handler, args, dueAt, attempts, error, createdAt, updatedAt}`.
    pub records: Collection,
    pub handlers: &'static [(&'static str, &'static Method)],
    /// Failed attempts before a timer stays failed.
    pub max_attempts: f64,
    /// First retry delay in milliseconds, doubling per failure.
    pub retry_delay_ms: f64,
    pub max_retry_delay_ms: f64,
}

pub(crate) fn integer(value: f64, label: &str, minimum: f64) -> Result<()> {
    if !is_safe_integer(value) || value < minimum {
        return type_error(format!(
            "{label} must be a safe integer at least {}",
            json::number(minimum)
        ));
    }
    Ok(())
}

pub(crate) fn require_name(value: &str, label: &str) -> Result<()> {
    if value.is_empty() {
        return type_error(format!("{label} must be a nonempty string"));
    }
    Ok(())
}

impl Scheduler {
    pub const fn new(
        name: &'static str,
        handlers: &'static [(&'static str, &'static Method)],
    ) -> Self {
        Scheduler {
            name,
            records: Collection::new(name).indexes(INDEXES),
            handlers,
            max_attempts: 3.0,
            retry_delay_ms: 1_000.0,
            max_retry_delay_ms: 60_000.0,
        }
    }
    pub const fn retries(
        self,
        max_attempts: f64,
        retry_delay_ms: f64,
        max_retry_delay_ms: f64,
    ) -> Self {
        Scheduler {
            max_attempts,
            retry_delay_ms,
            max_retry_delay_ms,
            ..self
        }
    }

    fn clock(&self, ctx: &mut Ctx) -> Result<f64> {
        let now = ctx.now()?;
        integer(now, "Server time", 0.0)?;
        Ok(now)
    }

    fn handler(&self, alias: &str) -> Result<&'static Method> {
        match self.handlers.iter().find(|(name, _)| *name == alias) {
            Some((_, method)) => Ok(method),
            None => fail(
                "SCHEDULER_HANDLER_MISSING",
                format!("Unknown scheduler handler {}", json::string(alias)),
            ),
        }
    }

    fn timer(id: &Value, stored: &Value) -> Value {
        object! {"id" => id}.with(stored)
    }

    pub fn get(&self, ctx: &mut Ctx, id: &str) -> Result<Value> {
        require_name(id, "Timer ID")?;
        let timer = ctx.get(&self.records, &Value::from(id))?;
        Ok(if timer.is_null() {
            Value::Null
        } else {
            Self::timer(&Value::from(id), &timer)
        })
    }

    fn first(&self, ctx: &mut Ctx, bound: Option<f64>) -> Result<Value> {
        let range = Range {
            prefix: vec![Value::from("pending")],
            lte: bound.map(Value::from),
            limit: 1,
            ..Range::default()
        };
        let page = ctx.range(&self.records.by("due").range(range))?;
        Ok(page
            .rows
            .into_iter()
            .next()
            .map_or(Value::Null, |row| Self::timer(&row.key, &row.value)))
    }

    fn schedule(
        &self,
        ctx: &mut Ctx,
        id: &str,
        due_at: f64,
        handler: &str,
        args: Value,
    ) -> Result<Value> {
        require_name(id, "Timer ID")?;
        integer(due_at, "Deadline", 0.0)?;
        if let Some(schema) = self.handler(handler)?.args
            && let Err(error) = schema.parse(&args)
        {
            return fail_with(
                "INVALID_ARGUMENT",
                format!("{handler} arguments: {}", error.message()),
                object! {"path" => error.path_value()},
            );
        }
        let time = self.clock(ctx)?;
        let key = Value::from(id);
        let previous = ctx.get(&self.records, &key)?;
        let created_at = previous.number("createdAt").unwrap_or(time);
        let timer = object! {
            "state" => "pending", "handler" => handler, "args" => args, "dueAt" => due_at, "attempts" => 0,
            "error" => Value::Null, "createdAt" => created_at, "updatedAt" => time,
        };
        ctx.set(&self.records, &key, timer.clone())?;
        Ok(Self::timer(&key, &timer))
    }

    /// Run `handler(args)` after `delay_ms`. Replacing an ID debounces earlier
    /// work and starts a new retry budget.
    pub fn after(
        &self,
        ctx: &mut Ctx,
        id: &str,
        delay_ms: f64,
        handler: &str,
        args: Value,
    ) -> Result<Value> {
        integer(delay_ms, "Delay", 0.0)?;
        let now = self.clock(ctx)?;
        self.schedule(ctx, id, now + delay_ms, handler, args)
    }

    pub fn at(
        &self,
        ctx: &mut Ctx,
        id: &str,
        due_at: f64,
        handler: &str,
        args: Value,
    ) -> Result<Value> {
        self.schedule(ctx, id, due_at, handler, args)
    }

    pub fn cancel(&self, ctx: &mut Ctx, id: &str) -> Result<bool> {
        if self.get(ctx, id)?.is_null() {
            return Ok(false);
        }
        ctx.delete(&self.records, &Value::from(id))?;
        Ok(true)
    }

    /// Timers in deadline order, optionally of one state ("pending" or "failed").
    pub fn scan(&self, ctx: &mut Ctx, state: Option<&str>) -> Result<Vec<Value>> {
        let states = match state {
            None => vec!["pending", "failed"],
            Some(state @ ("pending" | "failed")) => vec![state],
            Some(_) => return type_error("Unknown timer state filter"),
        };
        let mut timers = Vec::new();
        for current in states {
            let mut after: Option<String> = None;
            loop {
                let range = Range {
                    prefix: vec![Value::from(current)],
                    limit: 64,
                    after: after.take(),
                    ..Range::default()
                };
                let page = ctx.range(&self.records.by("due").range(range))?;
                timers.extend(
                    page.rows
                        .iter()
                        .map(|row| Self::timer(&row.key, &row.value)),
                );
                match page.cursor {
                    Some(cursor) => after = Some(cursor),
                    None => break,
                }
            }
        }
        timers.sort_by(|a, b| {
            let (a_due, b_due) = (
                a.number("dueAt").unwrap_or(0.0),
                b.number("dueAt").unwrap_or(0.0),
            );
            a_due
                .partial_cmp(&b_due)
                .unwrap_or(core::cmp::Ordering::Equal)
                .then_with(|| {
                    json::compare(
                        a.text("id").unwrap_or_default(),
                        b.text("id").unwrap_or_default(),
                    )
                })
        });
        Ok(timers)
    }

    /// Requeue a failed timer with the deployed handler and a fresh retry budget.
    pub fn retry(&self, ctx: &mut Ctx, id: &str, delay_ms: f64) -> Result<Value> {
        integer(delay_ms, "Delay", 0.0)?;
        let timer = self.get(ctx, id)?;
        if timer.text("state") != Some("failed") {
            return fail("TIMER_NOT_FAILED", "Only failed timers can be retried");
        }
        self.handler(timer.text("handler").unwrap_or_default())?;
        let time = self.clock(ctx)?;
        let mut stored = timer.into_object().unwrap_or_default();
        stored.remove("id");
        let mut retried = Value::Object(stored);
        retried.set("state", "pending");
        retried.set("attempts", 0);
        retried.set("error", Value::Null);
        retried.set("dueAt", time + delay_ms);
        retried.set("updatedAt", time);
        ctx.set(&self.records, &Value::from(id), retried.clone())?;
        Ok(Self::timer(&Value::from(id), &retried))
    }
}

impl Task for Scheduler {
    fn name(&self) -> String {
        format!("scheduler:{}", self.name)
    }
    fn due(&self, ctx: &mut Ctx) -> Result<Option<f64>> {
        let this = self;
        Ok(this.first(ctx, None)?.number("dueAt"))
    }
    fn run(&self, ctx: &mut Ctx) -> Result<Value> {
        let this = self;
        let now = this.clock(ctx)?;
        let due = this.first(ctx, Some(now))?;
        if due.is_null() {
            return Ok(Value::Null);
        }
        // Removal is staged before dispatch; a handler may replace its own ID.
        ctx.delete(&this.records, due.get("id"))?;
        this.handler(due.text("handler").unwrap_or_default())?
            .compute(ctx, due.get("args").clone())?;
        Ok(object! {"timer" => due.get("id")})
    }
    fn on_error(&self, ctx: &mut Ctx, failure: &TaskFailure) -> Option<Result<Value>> {
        let this = self;
        Some((|| {
            let now = this.clock(ctx)?;
            let due = this.first(ctx, Some(now))?;
            if due.is_null() {
                return Ok(Value::Null);
            }
            let attempts = due.number("attempts").unwrap_or(0.0) + 1.0;
            let mut delay = this.retry_delay_ms;
            let mut attempt = 1.0;
            while attempt < attempts && delay < this.max_retry_delay_ms {
                delay = this.max_retry_delay_ms.min(delay * 2.0);
                attempt += 1.0;
            }
            let exhausted =
                attempts >= this.max_attempts || failure.failed_at > MAX_SAFE_INTEGER - delay;
            let id = due.get("id").clone();
            let mut previous = due.clone().into_object().unwrap_or_default();
            previous.remove("id");
            let error = &failure.error;
            let mut recorded =
                object! {"code" => error.get("code"), "message" => error.get("message")};
            if error.has("details") {
                recorded.set("details", error.get("details").clone());
            }
            let state = if exhausted { "failed" } else { "pending" };
            let mut timer = Value::Object(previous);
            timer.set("state", state);
            timer.set("attempts", attempts);
            timer.set(
                "dueAt",
                if exhausted {
                    due.get("dueAt").clone()
                } else {
                    Value::from(failure.failed_at + delay)
                },
            );
            timer.set("updatedAt", failure.failed_at);
            timer.set("error", recorded);
            ctx.set(&this.records, &id, timer)?;
            Ok(object! {"timer" => id, "state" => state, "attempts" => attempts})
        })())
    }
}

impl Component for Scheduler {
    fn collections(&'static self) -> Vec<&'static Collection> {
        vec![&self.records]
    }
    fn tasks(&'static self) -> Vec<&'static dyn Task> {
        vec![self]
    }
}
